src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleKeyCache.scala

```diff
@@ -178,25 +268,39 @@ class GoogleKeyCache(
       )).compile.drain
     } yield decodedKey
 
-  private def isKeyActive(mostRecentKey: GcsObjectName, serviceAccountKeys: List[ServiceAccountKey]): Boolean = {
-    val keyRetired = System.currentTimeMillis() - mostRecentKey.timeCreated.toEpochMilli > 86400000L * googleServicesConfig.googleKeyCacheConfig.activeKeyMaxAge
-
-    val keyPathPattern(project, petSaEmail, keyId) = mostRecentKey.value
+  /** Cleans up keys by deleting orphans and bringing total key count down to 9.
+    */
+  private def cleanupKeys(pet: PetServiceAccount, cachedKeyObjects: List[CachedKey], serviceAccountKeys: List[ServiceAccountKey]): IO[Unit] = {
+    // if 10 or more keys, delete until we are below 10.
+    val tooManyFunction: IO[_] = if (cachedKeyObjects.size >= 10) {
+      val numToDelete = cachedKeyObjects.size - 9
+      val toDelete = cachedKeyObjects.sortBy(_.timeCreated).takeRight(numToDelete)
+      logger.info(
+        s"cleanupKeys: ${pet.id.project.value}-${pet.serviceAccount.subjectId.value} is reaping ${toDelete.size} keys due to quota"
+      )
+      toDelete.parTraverse(cachedKey => removeKey(pet, cachedKey.keyId))
+    } else {
+      IO.unit
+    }
 
-    // The key may exist in the Google bucket cache, but could have been deleted from the SA directly
-    val keyExistsForSA = serviceAccountKeys.exists(_.id.value.contentEquals(keyId))
+    // if any keys in IAM which are not in cache, delete those orphans.
+    val cachedKeyIds: Set[ServiceAccountKeyId] = cachedKeyObjects.map(_.keyId).toSet
+    val unknownKeyIds: Set[ServiceAccountKeyId] = serviceAccountKeys.map(_.id).toSet -- cachedKeyIds
 
-    !keyRetired && keyExistsForSA
-  }
+    val orphansFunction: IO[_] = if (unknownKeyIds.nonEmpty) {
+      logger.info(
+        s"cleanupKeys: ${pet.id.project.value}-${pet.serviceAccount.subjectId.value} is reaping ${unknownKeyIds.size} orphaned keys"
+      )
+      unknownKeyIds.toList.parTraverse(keyId => IO.fromFuture(IO(googleIamDAO.removeServiceAccountKey(pet.id.project, pet.serviceAccount.email, keyId))))
+    } else {
+      IO.unit
+    }
 
-  private def cleanupUnknownKeys(pet: PetServiceAccount, cachedKeyObjects: List[GcsObjectName], serviceAccountKeys: List[ServiceAccountKey]): Future[Unit] = {
-    val cachedKeyIds = cachedKeyObjects.map(_.value).collect { case keyPathPattern(_, _, keyId) => ServiceAccountKeyId(keyId) }
-    val unknownKeyIds: Set[ServiceAccountKeyId] = serviceAccountKeys.map(_.id).toSet -- cachedKeyIds.toSet
+    for {
+      _ <- tooManyFunction
+      _ <- orphansFunction
```

Comment:
I would do orphans first since that may bring it down below 10

Reply:
I found this section tricky. First, it counts the number of keys _in cache_ and reduces that below 9. Then, it looks for any keys _in google_ which aren't in the cache and deletes those.

I really wanted to use the count from cache for the first step, since the count from Google is eventually-consistent. I realize the possibility we'd delete more than necessary in corner cases. given this context, do you think this is worth changing?

Reply:
sound good, no change